File: chp3_stacks_and_queues/q2.py
class Stack(object):
    def __init__(self):
        self.size = 300
        self.stack = [None] * self.size
        self.top = -1

    def push(self, value):
        if self.top > -1:
            self.top += 1
            if value < self.stack[self.top-1].min:
                self.stack[self.top] = StackNode(value, value)
            else:
                self.stack[self.top] = StackNode(value, self.stack[self.top-1].min)
        else:
            self.top += 1
            self.stack[self.top] = StackNode(value, value)

    def pop(self):
        if self.top > -1:
            result = self.stack[self.top]
            self.top -= 1
            return result

class StackNode(object):
    def __init__(self, value, min):
        self.value = value
        self.min = min


a = Stack()
a.push(100)
a.push(200)
a.push(300)
a.push(600)
a.push(400)
a.push(10)
a.push(1000)
a.push(5)

print(a.pop().min)
print(a.pop().min)
print(a.pop().min)
print(a.pop().min)
print(a.pop().value)

File: chp4_trees_and_graphs/q2/q2.cpp
#include <chrono>
#include <functional>
#include <iostream>
#include <random>
#include <stack>
#include <unordered_set>
#include <vector>
using std::cout;
using std::chrono::system_clock;
using std::default_random_engine;
using std::uniform_int_distribution;
using std::unordered_set;
using std::vector;

struct Node {
	Node(int i) { id = i; }
	int id;
	vector<Node*>	paths;
	unordered_set<int> paths_set;
};

bool IsRoute(Node *, Node *);

int main() {
	unsigned seed = system_clock::now().time_since_epoch().count();
	default_random_engine gen(seed);
	uniform_int_distribution<int> dist_paths(0, 3), dist_nodes(0, 49);

	vector<Node> graph;
	for (int i = 0; i < 50; ++i) {
		graph.push_back(Node(i));
	}

	for (auto &node : graph) {
		int num_paths = dist_paths(gen);
		for (int i = 0; i < num_paths; ++i) {
			int id = dist_nodes(gen);
			if (node.paths_set.find(id) == node.paths_set.end() 
					&& node.id != id) {
				node.paths.push_back(&graph[id]);
			}
		}
	}

	for (int i = 0; i < 50; ++i) {
		Node *start = &graph[dist_nodes(gen)];
		Node *goal = &graph[dist_nodes(gen)];
		cout << IsRoute(start, goal) << " ";
	} cout << "\n\n";

	return 0;
}

bool IsRoute(Node *start, Node *goal) {
	std::stack<Node*> stk;
	std::unordered_set<Node *> visited;
	stk.push(start);
	while (!stk.empty()) {
		Node *current = stk.top(); stk.pop();
		visited.insert(current);
		if (current->id == goal->id) {
			return true;
		} else {
			for (auto node : current->paths) {
				if (visited.find(node) == visited.end())
					stk.push(node);
			}
		}
	}
	return false;
}
